### PR TITLE
New documentation box design

### PR DESCRIPTION
**Pull request by OceanOak**
Changelog:

```
Internal improvements
-  Convert the documentation box to use Tailwind CSS.
-  Create a new documentation box design.
```
These are the first steps towards a new design for the documentation box.

Changes:
- Add function name and deprecated status to the documentation box.
- Display on error rail inside a tooltip.
- Create a new design for Type error + add Expected/Actual values to it.
- Use Tailwind CSS.

### Before -> After

**Function name + deprecated status**
Normal fn

**Deprecated:**
Frame 21

**On error rail:**
onErrorRail

**Type error:**
TypeError

### These are just close-ups to help determine if any styling or spacing changes are needed.

**On error rail:** 
Screenshot (11)

**Non-deprecated functions** 
Screenshot (10)

**Deprecated functions + Type error**
Screenshot (7)

### CODE REVIEW

**Comment by pbiggar** [comment]
> Should the new Tooltip be moved from settingsComponent? (It is not part of settings so it doesn't make sense to keep it there). If so, where should we put it?

Yes, I'd love to have a Tooltip component that we could use anywhere (I tried to make one for the side bar but struggled with HTML layout issues).

I guess since it's a component, it should go in client/src/components.

**Comment by OceanOak** [comment]
I've created a new tooltip that could be styled by the provider ( body, position, color, etc.)
1. Should we remove the one used in SettingsContributingView and use this one everywhere.
2. There is a file named Tooltips.res, it is used in the function space. Should we move the newly created tooltip there or should keep it in a seperate file?

**Comment by pbiggar** [comment]
> **Questions:**
> 
> 1. I've created a new tooltip that could be styled by the provider ( body, position, color, etc.). Should we remove the one used in SettingsContributingView and use this one everywhere?

If it has some useful settings that are reused in Settings, then having it could be useful to reduce duplication. If not, then redoing that tooltip to use the component is valuable so we have less concepts/code to keep track of.


> 2. There is a file named Tooltips.res, it is used in the function space and the welcome to dark tutorial. Should we move the newly created tooltip there or should keep it in a seperate file?

Keep it separate I think. Tooltips.res is mostly about a tutorial, so we could rename it to Tutorial.res perhaps?

> 3. After converting PrettyDocs.res to use the Tailwind, I encountered some failing tests in TestPrettyDocs.res. The failing tests all involve a class name being expected, but a Tailwind class is returned instead. To resolve this issue, I directly used the Tailwind classes in the tests. However, I am uncertain if this is the most appropriate solution. If there is a better way to fix it, please let me know.
> 
> Example:
> 
> ```
> Expected : var  // a class name
> Actual : text-purple1 //tailwind
> ```

That feels like a good approach. It is possibly one of the downsides to tailwind, but it's probably not terrible.


> **Issues:**
> 
> 1. On error rail appears only when cursor is on complete function name, when it used to appear on AC and when the cursor is on the arguments as well.
>    Previously we've used `list{ViewErrorRailDoc.hintForFunction(f, sendToRail)}` to display it. Now because we moved it inside a tooltip I had to test if the function is on Rail or NoRail in order to display `(On Error Rail "i")`. Is there a better way to do it?

Do you feel it should appear on AC and when the cursor is on the arguments? If so, then we should keep it. I don't quite understand the code issue - presumably you can pass the errorrail status from the function to the tooltip?

> 2. `fluid_show_docs_for_command_on_selected_code` integration test is failing. Didn't manage to find a way to fix it any guidence here is appriciated.

Since there wasn't a video or a trace.zip stored in CI, the best approach is to:
- try it locally, but change the playwright config to always save a video or trace that you can replay.
- manually repeat the steps from the integration test to see if that finds the flaw 
- load the canvas locally after running the integration test to see its state (go to darklang.localhost:9000/a/test-WHATEVER-THE-INTEGRATION-TEST-NAME-IS

**Comment by OceanOak** [comment]
> Do you feel it should appear on AC and when the cursor is on the arguments? If so, then we should keep it. I don't quite understand the code issue - presumably you can pass the errorrail status from the function to the tooltip?

I think it should when the cursor is on the arguments at least, because we rarely have the cursor on the function's name.

> Since there wasn't a video or a trace.zip stored in CI, the best approach is to:
> try it locally, but change the playwright config to always save a video or trace that you can replay.
> manually repeat the steps from the integration test to see if that finds the flaw
> load the canvas locally after running the integration test to see its state (go to darklang.localhost:9000/a/test-WHATEVER-THE-INTEGRATION-TEST-NAME-IS

This time, I didn't have to go through all of these steps,  but I still need to figure out how to run the integration tests because I can't run them inside the container or locally.

The fix was pretty straight forward I was just looking in the wrong place.
The issue is in `await page.waitForSelector(".documentation-box")` , which we no longer use. I commented it out to make sure that it was causing the issue. 

The question now is whether I should delete it or replace it with all of the Tailwind classes.

**Comment by pbiggar** [comment]
> The question now is whether I should delete it or replace it with all of the Tailwind classes.

I think it's fine to leave in semantic classes for things like this (it also makes the html more readable when we are looking at it in the browser dev tools). Just so long as we don't have CSS attached to it.

**Comment by OceanOak** [comment]
> I think it's fine to leave in semantic classes for things like this (it also makes the html more readable when we are looking at it in the browser dev tools). Just so long as we don't have CSS attached to it.

I couldn't use the Tailwind classes (I got issues with escaping some characters), so I've added an ID to the documentation box instead.

**Comment by OceanOak** [comment]
> Do you feel it should appear on AC and when the cursor is on the arguments? If so, then we should keep it. I don't quite understand the code issue - presumably you can pass the errorrail status from the function to the tooltip?

To explain this better:
1. If we don't know if the function is on the rail, we display a generic message (this is what we display when the cursor is on the argument).
2. If we know if the function goes on the rail, we display a specific message.

I managed to make it appear when the cursor is on the argument. The problem is when the function isn't a function that goes to the error rail I am still getting an empty tooltip. Here is an example: 
dark-Dark-Google-Chrome-2023-01

**Comment by pbiggar** [comment]
> I managed to make it appear when the cursor is on the argument. The problem is when the function isn't a function that goes to the error rail I am still getting an empty tooltip only when the cursor is on the argument. 

It shouldn't show the `(on ErrorRail (i))` there at all, since it's not on the rail.

**Comment by OceanOak** [comment]
> It shouldn't show the `(on ErrorRail (i))` there at all, since it's not on the rail.
Yes, that's what I trying to achieve. and it is not showing when the cursor is on the function's name only when it is on the argument.

**Comment by OceanOak** [comment]
Changes:

**Main.res** and **Tutorial.res** and **Entry.res** and **ViewUserFunction**
- Changed Tooltips to Tutorial because we renamed the Tooltips.res file to Tutorial.res 

**View.Res**
-converted the documentation box to Tailwind.
-added an ID for the documentation box to make the it easier to reference in the integration test
- changed Tooltips to Tutorial (because of the renaming)

**ViewErrorRailDoc.res**
-converted the error rail text that appears inside the tooltip to Tailwind and extracted the shared styles to avoid repeating the same code.

**SettingsCanvasesView.res** and **SettingsInviteView.res** and **SettingsView.res**
No changes in this file just formatting.

**SettingsViewComponents**
Fixed a comments and formatting.

**Tooltip.res**
A new file where I attempted to create a generic tooltip that we can use anywhere and can be styled by the provider (position, color,etc)

**FluidAutocomplete.res**
1. switched it to tailwind + changed the font 
2. type error
      - added expected/actual 
      - added a div that includes expected/actual and the error message.
      - extracted shared styles
      
image


3. added a documentation header deprecationHeader that specifies whether the function is deprecated or not and the functions name
image
image

If the function is deprecated we've added a tooltip with a generic text about function's being deprecated (it is kind of ugly but I don't have an idea how to make it better yet).
image

5. on error rail
moved on errorRail inside a tooltip
image

6. deprecationFooter
changed the styling
image

7. Function's return type
-added an icon
image

**PrettyDocs.res**
-switched to tailwind 

**TestPrettyDocs.res**
-used tailwind instead of css class

**Tailwind.config.js**
-expanded the color palette + put the fontsizes in the right order

File: client/src/fluid/FluidAutocomplete.res [hunk 1]
```diff
@@ -763,55 +812,131 @@ let documentationForFunction = (
...
+    let sharedStyle = %twc("font-text font-medium text-xs tracking-wide")
     let deprecationFooterContents = switch f.deprecation {
     | NotDeprecated => list{}
-    | ReplacedBy(name) => list{Html.text("replaced by " ++ FQFnName.StdlibFnName.toString(name))}
-    | RenamedTo(name) => list{Html.text("renamed to " ++ FQFnName.StdlibFnName.toString(name))}
-    | DeprecatedBecause(reason) => list{Html.text(reason)}
+    | ReplacedBy(name) => list{
+        Html.span(
+          list{tw(sharedStyle)},
+          list{Html.text("replaced by " ++ FQFnName.StdlibFnName.toString(name))},
```
Comment (pbiggar) [inline client/src/fluid/FluidAutocomplete.res:906]: Can you style the function name here?

File: client/src/fluid/FluidAutocomplete.res [hunk 2]
```diff
@@ -763,55 +812,131 @@ let documentationForFunction = (
...
+        ),
+      }
+    | RenamedTo(name) => list{
+        Html.span(
+          list{tw(sharedStyle)},
+          list{Html.text("renamed to " ++ FQFnName.StdlibFnName.toString(name))},
```
Comment (pbiggar) [inline client/src/fluid/FluidAutocomplete.res:912]: Can you style the function name here?

File: client/src/fluid/FluidAutocomplete.res [hunk 3]
```diff
@@ -707,7 +710,11 @@ let selectUp = (a: t): t =>
 
 let isOpened = (ac: t): bool => Option.isSome(ac.index)
 
+let fontSharedStyle = %twc("font-text tracking-wider")
+
 let typeErrorDoc = ({item, validity}: data): Vdom.t<AppTypes.msg> => {
+  let expected = Html.span(list{tw2(fontSharedStyle, %twc("mx-1"))}, list{Html.text("Expected: ")})
```
Comment (pbiggar) [inline client/src/fluid/FluidAutocomplete.res:716]: Having the green/orange text have the same foreground as background is a little jarring - can it be changed?

Also, it strikes me that this green/orange isn't from our main color palate - should it be changed to colors from our palate?

File: client/src/ui/PrettyDocs.res [hunk 4]
```diff
@@ -12,12 +12,19 @@ let codeEx = "^(.*)\\{\\{(.+)\\}\\}(.*)$"
 
 let linkEx = "^(.*)\\[(.+)\\]\\((http[s]?\\://.+)\\)(.*)$"
 
-let codeClass = "code"
-
 let nestedTag = Regex.regex(`<\\\w+[^>]*<`)
 
 let nestedCodeBlock = Regex.regex(`{{[^}]+{{`)
 
+let \"type" = %twc("text-green")
```
Comment (pbiggar) [inline client/src/ui/PrettyDocs.res:19]: You can calling this variable `typ` instead.

File: tailwind.config.js [hunk 5]
```diff
@@ -27,11 +27,14 @@ module.exports = {
 
       red: "#ab4642",
       orange: "#dc9656",
+      orange1: "#a67254",
```
Comment (pbiggar) [inline tailwind.config.js:30]: Can you add a comment for orange1 like you did for purple1?

**Review by pbiggar — APPROVED** [review]
This all looks great. I have some comments but please address them as a followup PR so that this can be merged ASAP.

**Pull request merged**